Fix crash when lowering decision dag

RikkiGibson opened the pull request:
Fixes #46536 based on @jcouv's suggestion.

/cc @cston @jaredpar @gafter

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs:
```diff
@@ -5116,6 +5116,368 @@ .maxstack  2
+            (_, 2) => 2,
+            (_, 3) => 3,
+            (_, 4) => 4,
+            _ => 5
```
I think this is a scenario where we'd want a second `switch` to be generated (one for the case where `x` is `""`, and one for the other cases). Can you file an issue on that?

RikkiGibson replied:
Opened #47145

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs:
```diff
@@ -5116,6 +5116,368 @@ .maxstack  2
+            var verifier = CompileAndVerify(source, expectedOutput: "0123233", options: TestOptions.DebugExe);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("Program.M", @"{
```
I see that we didn't generate a switch dispatch in this scenario, so this looks like a different scenario than the one we looked at together. Could you explain what was happening in this case?

RikkiGibson replied:
We lower into a BoundSwitchDispatch, so the LocalRewriter code path we're interested in is exercised, but the BoundSwitchDispatch [is not sufficiently complex](https://github.com/dotnet/roslyn/blob/892f9fae716ffb7b6fbd4a32db1eeca8fd376f6f/src/Compilers/Core/Portable/CodeGen/SwitchStringJumpTableEmitter.cs#L225) for the compiler to decide to switch on the string hash code and it instead emits an if/else if/... chain.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs:
```diff
@@ -5416,7 +5416,7 @@ static int M(string x, string y)
 }
 ";
-            var verifier = CompileAndVerify(source, expectedOutput: "0123233", options: TestOptions.DebugExe);
+            var verifier = CompileAndVerify(source, expectedOutput: "0123233", options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
```
Why was the language version necessary? C# 9 should be the default at this point (we shouldn't be using "preview" until we add new features)

RikkiGibson replied:
not the case in 16.7 branch.

RikkiGibson commented:
@jcouv I believe I have addressed all your comments. Please let me know if you have any others.

@jaredpar Please let me know if you need anything from me to help drive the approval process.

jcouv commented:
Tagging @gafter again as FYI (consider reviewing this fix after the fact), and to comment on follow-up issue https://github.com/dotnet/roslyn/issues/47145
